### ISSUE TITLE
[WIP] Compositional Router API [RFC]

### ISSUE THREAD

**Issue by jwx** [aurelia/aurelia#624]
- [x] Compositional routing
- [x] Navigation instructions, `NavigationInstruction` and `ViewportInstruction`
- [x] Navigating with `goto` and `a`nchor links
- [x] No route configuration needed
- [x] Manually configure v1 routes
- [x] Manually configure v1 child routes

# 💬 RFC

This RFC will describe how to manually configure routes in code in the new Aurelia 2 router. Manually configuring routes in code is an addition to, not a replacement of, the dynamic routing and routes configured with decorators.

# Compositional routing

The Aurelia 2 router is completely new with a new principle: based on one or more **navigation instructions** it creates compositions of components in viewports. Unlike the router of Aurelia v1, the new router isn't reading the browser Location url to get the information about the full page content. Instead it accepts a set of navigation instructions - viewport/component pairs - through links or code and then, once navigation is completed, it updates the browser Location url to reflect the total composition. In order to accurately represent this compositioned state, the new router has expanded the url syntax of the Aurelia v1 router.

With the Aurelia v1 router the browser Location url is central and strategies and configuration for how the url string describes/maps page content is necessary. With the new Aurelia 2 router, navigation instructions directly describe which components to show in what viewports.

# Navigation instructions

A navigation instruction consists of three main parts: the **component** to load, the **parameters** to pass into it when loading it and the **viewport** to load it in.
```typescript
{ component: 'product', parameters: { id: 'abc123' }, viewport: 'main-content' }
```
If there are no parameters to pass into the component, `parameters` can be omitted. If the router can deduce which viewport is intended -- either through viewport configuration or availability; there's only one available viewport for the component -- `viewport` can also be omitted.

## `NavigationInstruction`

The main type for the navigation instructions is the `NavigationInstruction` type. It's a type union that aims to provide a simple and easy to use way to instruct the router. As seen above, it can be a plain object with the following properties
- `component`: a component name (`'product'`), type (`Product`) or instance (`myProductVM`)
- `parameters`: an object (`{ id: 'abc' }`), an array (`['abc']`) or a string (`'id=abc'` or `'abc'`)
- viewport  a viewport name (`'main-content'`) or instance (`myMainViewportProperty`) 

It can also be a string containing the string versions of the above properties formatted as `component(parameters)@viewport`. Values for `parameters` and `viewport` may be omitted according to above. For example
```typescript
'product(id=abc,lang=sv-SE)@main-content'
'product(abc,sv-SE)@main-content'
'product(abc)'
'products'
```

## `ViewportInstruction`

The router translates incoming `NavigationInstruction`s into `ViewportInstruction`s, the type of object the router uses internally to represent navigation instructions. `ViewportInstruction` is however also a part of the type union `NavigationInstruction`, meaning that
```typescript
new ViewportInstruction(component, parameters, viewport)
```
is also a valid navigation instruction. The constructor parameters follows the same principles as the plain object properties of the `NavigationInstruction` so `component` can be a name, a type or an instance, `parameters` can be an object, an array or a string, `viewport` can be a name or an instance and both `parameters` and `viewport` are optional.

Most of the router API will accept `NavigationInstruction`s as navigation instructions, but there are some lower level APIs that will only accept `ViewportInstruction`s.

# Navigating

There are two main ways to instruct the Aurelia 2 router to navigate, one in code and one in html. To navigate in code inject `IRouter` into the component (or service or any other code object) and call its `goto` method with the navigation instruction
```typescript
this.router.goto({ component: 'product', parameters: { id: 'abc' }, viewport: 'main-content' });
this.router.goto({ component: Products, parameters: { filter: variableWithFilter } });
this.router.goto('product(abc)');
this.router.goto(`products(${variableWithFilter})`); 
```
To navigate in html, add an `a`nchor link with the custom attribute `goto`
```html
<a goto="product(id=abc)@main-content">Fully qualified</a>
<a goto="product(abc)">Same as previous when router can deduce viewport</a>
<a goto="products(${variableWithFilter}">Standard string interpolation works</a>
<a goto.bind="{ component: 'product', parameters: { id: 'abc' }, viewport: 'main-content' }">Fully qualified</a>
<a goto.bind="{ component: 'products', parameters: { filter: variableWithFilter } }">When router can deduce viewport</a>
```
The similarities between the html and the code is no coincidence; behind the scenes the `goto` custom attribute is handled by the `goto` method.

## `goto` options

The router's `goto` method is more or less the single interface for navigating. In order to support additonal use cases, it accepts a second parameter with `options`
```typescript
this.router.goto('product(abc)', { replace: true }); // Replace current history entry
this.router.goto('notes@sidebar', { append: true }); // Append an additional viewport instruction to current navigation, useful in router lifecycle methods such as `enter`
```
The documention for `goto` provides a detailed list of the available options.

## Navigating in multiple viewports at the same time

Using more than one viewport is a good way to use composition in the UI. For example, with two sibling viewports, `sidebar` and `main-content`, navigation can occur individually within those two viewports without the other viewport having to be involved or even aware.
```html
<div> <!-- sidebar of the screen -->
  <a goto="sidebar1">Sidebar content 1</a>
  <a goto="sidebar2">Sidebar content 2</a>
  <au-viewport name="sidebar" used-by="sidebar1,sidebar2"></au-viewport>
</div>
<div> <!-- most of the screen -->
  <a goto="main1">Main content 1</a>
  <a goto="main2">Main content 2</a>
  <au-viewport name="main-content"></au-viewport>
</div>
```
Another useful structure is parent-child viewports
`main.html`
```html
<a goto="product(abc)">Product abc</a>
<au-viewport name="main"></au-viewport>
```
`product.html`
```html
<a goto="product-suppliers(${id})">Suppliers for ${name}</a>
<a goto="product-manuals(${id})">Manuals for ${name}</a>
<au-viewport name="product-details" default="product-suppliers(${id})" used-by="product-suppliers,product-manuals"></au-viewport>
```
In this case, navigations in `main.html` naturally affects the `product-details` viewport since `main` can load a different component than `product`. Navigations in `product.html` however does not affect or even need to be aware of the `main` viewport. Furthermore, `main.html`, and the app above it, does not need to be aware of or take any actions to facilitate navigations within `product.html`. (These possibilites for encapsulation are quite useful for plugins.)

Both of the above examples have navigations involving a single viewport at a time, but it's also possible to perform simultaneous navigations in more than one viewport. The router's `goto` method accepts either a `NavigationInstruction` or an array of `NavigationInstruction`s as first parameter making the following possible
```typescript
this.router.goto([ { component: sidebar2 }, 'main2' ]);
this.router.goto([ 'product(${id})', { component: ProductManuals, parameters: { id: id } } ]);
```
Note that when dealing with sibling viewports the order between the navigation instructions doesn't matter but when dealing with parent-child viewports the parent instruction needs to come before the child instruction.

The two navigations above can also be achieved with the string syntax
```typescript
this.router.goto('sidebar2+main2');
this.router.goto('product(${id})/product-manuals(id=${id})');
```
and
```html
<a goto="sidebar2+main2">Load sidebar2 and main2</a>
<a goto="product(${id})/product-manuals(id=${id})">Load product ${id} and its manuals</a>
```
Note that in the string syntax siblings are separated with `+` and parent-children with `/`. 

`goto.bind` also accepts an array of `NavigationInstruction`s
```html
<a goto.bind="[ { component: sidebar2 }, 'main2' ]">Load sidebar2 and main2</a>
<a goto.bind="[ 'product(${id})', { component: 'product-manuals', parameters: { id: id } } ]">Load product ${id} and its manuals</a>
```
(Also note that the examples above mixes types of `NavigationInstruction`s to illustrate that it's possible and not because it's necessarily a good thing to do.)

# No route configuration needed

In Aurelia 2 all components need to be registered with DI before being used. This is a general requirement that has nothing to do with the router, but it's something the router can take advantage of. As a result, there's no need for any route configuration for compositional routing to work. Once the `RouterConfiguration` has been registered at Aurelia start up all components become routeable when they are registered. (This behavior can be changed so that components need to be marked as routeable before the router can load them in viewports *(not implemented yet)*.) Components can also be made routeable on more than their registered name by registering aliases in DI. *(Not implemented yet, awaiting DI aliasing PR.)*

All the examples above, including sibling and parent-child viewports, works without route configuration.

# Manual configuration

While not necessary, the Aurelia 2 router provides a way to manually configure routes following the same path patterns as v1 *(not implemented yet)*. To manually register a route inject `IRouteConfiguration` into a component (or service or any other code object) and call its `register` method with the path pattern to match and the navigation instruction(s) to perform if it's matched

```typescript
this.routeConfiguration.register('/product/:id', { component: 'product', viewport: 'main-content' });
this.routeConfiguration.register('/products', { component: Products, parameters: { filter: variableWithFilter } });
this.routeConfiguration.register('/both2', [ 'sidebar2', 'main2' ]);
this.routeConfiguration.register('/product-with-manuals/:name', 'product(:name)/product-manuals(id=:name');
```
All matched `:parameters` will be either transfered with the pattern name (first example) or mapped against `:` specified locations (fourth example) into each of the `NavigationInstruction`(s).

## Manual configuration of more than one route

For convenience, and the possibility to easily feed json data to the route configuration, the first parameter of `register` can also be an array of route configurations
```typescript
this.routeConfiguration.register([ 
    { path: '/products', instructions: { component: 'products' } },
    { path: '/product-with-manuals/:name', instructions: [ 'product', 'product-manuals' ] },
]);
```

## `register` options

There will also be a third parameter/property to `register`, `options`, that will contain configuration options that's neither the path or the instructions
```typescript
this.routeConfiguration.register(
    '/product/:id', 
    { component: 'product', viewport: 'main-content' }, 
    { transformPath: true }, // Transform configured url path to compositional instructions in browser Location url, defaults to false
);
this.routeConfiguration.register([ 
    { path: '/products', instructions: { component: 'products' }, options: { transformPath: true } },
    { path: '/product-with-manuals/:name', instructions: [ 'product', 'product-manuals' ], options: { transformPath: true } },
]);
```
The documention for `register` provides a detailed list of the available options.

**Comment by davismj** [comment]
examples?

**Comment by fkleuver** [comment]
@davismj For context, we had a discussion about this yesterday. As you might recall I said a while ago that I would work on your RFCs eventually and do the integrations and, if necessary, make any architectural changes.

Your RFCs still have a number of open questions regarding details about underlying functionality such as lifecycle interactions, timings, etc, that probably require some experimentation to settle on.

Now as it turns out I've got more work left on the core and AOT than I anticipated I would have at this point, so I asked @jwx if he wanted to have a crack at providing an API that would lay the groundwork for backwards compatibility with v1 as well as your RFCs. In short, a configuration API.

So I explained broadly how the route recognizer works. The basic idea is that there is some kind of `registerRoute` method that accepts a pattern and a navigation instruction, much like `mapRoute` in v1.

And that's really all there is to it: a configuration api utilizing the route recognizer, laying the groundwork for your RFCs and v1 backwards compatibility.

**Comment by EisenbergEffect** [comment]
A couple quick comments.

* If possible, I'd prefer the actual router configuration to look more like Aurelia 1. For example, a route config should be a simple json object rather than a multi-parameter method call. We have plenty of scenarios where apps are actually dynamically generating the config on the server, and then they want to pass it straight to the router.
* Above you say: `component: a component name ('product'), type (Product) or instance (myProductVM)`. Have you tried getting this to work with an existing component instance? I think there are going to be lots of issues with that. I don't recommend trying to support that.

**Comment by jwx** [comment]
@EisenbergEffect You mean something like the below?
```typescript
this.routeConfiguration.register([ 
    { path: '/products', instructions: { component: 'products' } },
    { path: '/product-with-manuals/:name', instructions: [ 'product', 'product-manuals' ] },
]);
```
If so, that's more or less how I had intended the "register multiple paths at once" to look like. If not, could you please elaborate?

Yeah, I've got component instance for `component` working. No issues so far, but I need more tests to say for sure that it's working fully.

**Comment by EisenbergEffect** [comment]
Yep, that's what I was looking for. Thanks!

**Comment by EisenbergEffect** [comment]
In terms of route param syntax, I think we had some notes on how we wanted to change that so that we could support type conversion. @davismj may remember the details.

**Comment by davismj** [comment]
```ts
this.routeConfiguration.register([ 
    { path: '/products', instructions: { component: 'products' } },
    { path: '/product-with-manuals/:name', instructions: [ 'product', 'product-manuals' ] },
]);
```

Any reason add and remove routes couldn't be a method directly on the router?

**Comment by jwx** [comment]
@davismj None other than that I like to keep things separated when appropriate. And since none of the route configuration code or the configured routes will be a part of, or even understood by, the router itself it seemed like a good candidate to separate. But adding

**Comment by koenbeuk** [comment]
> it updates the browser Location url to reflect the total composition.

What would a url look like and how do multiple viewports affect it? How can we customize the generated url's.

In one use-case, I'm dealing with a site with 2 viewports: [main][sidebar], I may navigate to something like groups/foo/products in main, fetch a list of products and when clicking on one, a sidebar would come popping out with various tabs: 'details, orders, purchase history'... Here URL's are meant to  control what is rendered, be able to pass these URL's around (e.g. by email) and be able to navigate back into the exact link. So in this case, URL's need to take on the form of something in the lines of:

```
/groups/foo/products/1+/details (+ symbol, may be something else as a way of separating various instructions)
/groups/foo/products/2+/orders/2
```

Opening either link in a browser should restore both viewports in their desired state.

**Comment by jwx** [comment]
@koenbeuk There will be three main ways for the Location url/deep links to be set:
1. The compositional conventions can create a deep link with complete state. If more than one viewport is involved they'll indeed be separated by either `+` for siblings or `/` for parent-child viewports. (The characters and format of these generated deep links can be configured as well.)
2. Manually configured routes can be set to use the route path for the deep link (default behavior).
3. The router configuration has a property pair called `transformFromUrl`/`transformToUrl` that gives more or less full control over deep links. A simplified (and a bit silly) example:
```typescript
RouterConfiguration.customize({
  transformFromUrl: route => route === 'shortcut' ? 'long/list-of/routes+details' : route,
  transformToUrl: route => route === 'long/list-of/routes+details' ? 'shortcut' : route,
})
```

To sum things up, you won't _need_ to think about the Location url, but if you _want_ to, you certainly can.

**Comment by khuongduybui** [comment]
I really like the idea of a file-based router (as in svelte/sapper)
I don't think Aurelia 2 needs to ship with a file-based router, but at least the routing API should be flexible enough that I can build a file-based router plugin on top of it.

**Comment by EisenbergEffect** [comment]
@khuongduybui Can you go into a few more details as to what features you would want? "file-based router" is pretty broad.

**Comment by khuongduybui** [comment]
Implementation details are up for discussion, but conceptually, a file-based router solution looks at my source code directory and infer my routes from that. For example, please see https://github.com/jakobrosenberg/svelte-filerouter#file-scheme

**Comment by EisenbergEffect** [comment]
FYI That could be built on Aurelia 1 today without much work. All you have to do is generate a json file during your build step. I don't see any problem with Aurelia 2's router being able to support that based on our plans.

**Comment by fkleuver** [comment]
@khuongduybui In general we are trying to make sure that things that work at build time, also work at runtime. Svelte only works with the svelte compiler; normal svelte code won't do anything in native js (this manifests as: you can't really integrate svelte into a legacy app or existing cms where you typically can't do much with the build process)

For a file based router we could make this work at runtime with requirejs, but then you wouldn't be able to do stuff like automatically generate a menu. That would only work with restricting the feature to build time only. Would you, as a user, not have any problems with that restriction?

If that is the case, and more people feel the same, we could consider to vastly expand our conventions to include things like this but it would create an AOT-only feature set that's distinctly isolated. Again, it's something we're ourselves preferring to avoid because it creates a disparity where some features aren't available in JIT scenarios, which might upset users who want to use Aurelia to progressively enhance legacy apps. I'm not entirely sure though. It's not like we've really asked around..

**Comment by khuongduybui** [comment]
1. @EisenbergEffect yes, and in fact, that's the plugin I'm planning to write. I just want to make sure before I write the plugin that the concept of file-based routing is something Aurelia 2 can support by design, not just by a co-incidence.
2. @fkleuver you are correct that file-based routing is a build-time only feature. However, there is nothing stopping us to generate a build-time routing ruleset, which can be read / modified / replaced at runtime.

**Comment by EisenbergEffect** [comment]
@khuongduybui We'd certainly love to partner with you to see that delivered to the community. I think it should be quite possible to build this based on our current design. @fkleuver Do you have any thoughts on if/how a file-based routing "pass" could be integrated with AOT? I can imagine something like this wanting to inject `route` decorators or other similar types of config during the build process but before your optimizations are run. It's sort of like a specific conventions recipe actually.

**Comment by fkleuver** [comment]
That would be pretty trivial once AOT how I'm planning it is in place. You could just ask the AOT api "what are all potential links in the app and from which components do they appear" and "what are all the files in this project" and then use the route recognizer or some custom matcher logic to pair them up, then tell AOT "ok, these files are now routes".

So the only unknown is the "match potential links with potential file names" logic which I'm not sure if the route recognizer itself could do that in a straight-forward manner but given that that is implemented, it would be less than 10 LOC of aot api calls to get the other info. Would be using exactly that same kind of api for example for our vs code plugin to provide context on available variables, resources, etc in the html template

**Comment by khuongduybui** [comment]
Ok thanks champs. I think I've got enough to get started with a plug-in for v1. We'll cross that migration path bridge when we get to it.

**Comment by jwx** [comment]
Implemented.
